src/interface/sparse_pack_base.cpp

```diff
@@ -167,6 +167,64 @@ SparsePackBase SparsePackBase::Build(T *pmd, const PackDescriptor &desc) {
           }
         }
       }
+      for (auto &pv : pmbd->GetFaceVector()) {
+        if (desc.IncludeVariable(i, pv.get())) {
+          if (pv->IsAllocated()) {
+            for (int t = 0; t < pv->GetDim(6); ++t) {
+              for (int u = 0; u < pv->GetDim(5); ++u) {
+                for (int v = 0; v < pv->GetDim(4); ++v) {
+                  pack_h(1, b, idx) = pv->data.x1f.Get(t, u, v);
+                  if (ndim > 1) {
+                    pack_h(2, b, idx) = pv->data.x2f.Get(t, u, v);
+                  }
+                  if (ndim > 2) {
+                    pack_h(3, b, idx) = pv->data.x3f.Get(t, u, v);
+                  }
+                  idx++;
+                }
+              }
+            }
+          }
+        }
+      }
+      for (auto &pv : pmbd->GetEdgeVector()) {
+        if (desc.IncludeVariable(i, pv.get())) {
+          if (pv->IsAllocated()) {
+            for (int t = 0; t < pv->GetDim(6); ++t) {
+              for (int u = 0; u < pv->GetDim(5); ++u) {
+                for (int v = 0; v < pv->GetDim(4); ++v) {
+                  pack_h(1, b, idx) = pv->x1e.Get(t, u, v);
+                  if (ndim > 1) {
+                    pack_h(2, b, idx) = pv->x2e.Get(t, u, v);
+                  }
+                  if (ndim > 2) {
+                    pack_h(3, b, idx) = pv->x3e.Get(t, u, v);
+                  }
+                  idx++;
+                }
+              }
+            }
+          }
+        }
+      }
+
+      for (auto &pv : pmbd->GetNodeVector()) {
+        if (desc.IncludeVariable(i, pv.get())) {
+          if (pv->IsAllocated()) {
+            for (int t = 0; t < pv->GetDim(6); ++t) {
+              for (int u = 0; u < pv->GetDim(5); ++u) {
+                for (int v = 0; v < pv->GetDim(4); ++v) {
+                  pack_h(0, b, idx) = pv->data.Get(t, u, v);
+                  PARTHENON_REQUIRE(
+                      pack_h(0, b, idx).size() > 0,
+                      "Seems like this variable might not actually be allocated.");
+                  idx++;
+                }
+              }
+            }
+          }
+        }
+      }
```

Comment:
I don't love having to add all this essentially identical boilerplate. If we moved to all variables being the same type, i.e., make them not structs anymore, this could all go away.

Alternatively, if you're married to the struct machinery (not my preference), we could probably make some generic templated functions for doing this looping.

Reply:
I don't like this repeated looping either, nor would these be the only loops repeated. On the other hand, if we only keep the cell-centered loop, we'd need to make sure the loop applies equally well to face-centered and edge-centers variables.  Would one of the indices in the variable be dedicated then to 3-directions so that face-centered and edge-centered variables have one less dimension available? Would we also have to make sure that the coarse variable and flux variable are never allocated? Or only pack the dimensions as needed, like we do for fluxes?

Reply:
I'm not sure I fully understand the question. In principle, the loop is just flattening over dims 6,5,4, which should be generic over grid location. The reason you have special logic for faces and edges is because they live in structs. I think that could be resolved in various ways though.

Reply:
I agree that we want to avoid code repetition like this if at all possible. As I look at things, I am more and more thinking that the relatively simple, fairly flexible way to go is in `CellVariable` to switch
```c++
ParArrayND<T, VariableState> data; 
ParArrayND<T, VariableState> coarse_s;
ParArrayND<T, VariableState> data_all[8]; 
ParArrayND<T, VariableState> coarse_s[8]; 
// After allocation 
data = data_all[0];
coarse_s = coarse_s_all[0]; 
``` 
with `0->cell center`, `1->x face`, etc. and then only allocating the pieces which are necessary in `CellVariable::AllocateData()`. I think by pointing `data` and the cell centered value, it wouldn't break anything that is already in the code. We would then need to add an extra index to the sparse pack `operator()` specifying the cell element that is being requested and change the internal indexing a little bit.

Reply:
I spoke to @jdolence about this this afternoon and he pointed out that different parts of the data array may be one copy or not. For example cell-centered data for B fields may not need to have a separate copy per stage. To me this still suggests the treatment of face-centered fields being variables as variables are currently written, just with a different shape

Reply:
I think that fits fine within this model though, since it is flexible enough that you can either have the field on cell centers, faces, or on both. If you want different metadata for the cell centers and the faces, you just create two separate fields.

Reply:
But if that's the case, is there a reason to add the extra complexity of all these different arrays? I think having different metadata for different locations will be common.

Reply:
Of course it is subjective, but I think it introduces limited extra complexity. It is very similar to the amount of complexity required for possibly carrying around fluxes at the end of the day. I can also think of the example of geometric quantities for a static space time, which we would want at the cell centers and cell faces, and would be one copy everywhere.